teps/0015-pending-pipeline.md

```diff
@@ -0,0 +1,121 @@
+    - [Platform Implementor](#platform-implementor)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes a method for pausing `PipelineRuns` that are then ignored by
+the Tekton reconciler. These pending resources can be used by platform
+implementers to control when a `PipelineRun` can create new `TaskRuns`, for
+example, in cases where the cluster is under heavy load and not ready to process
+new `TaskRuns`.
+
+`PipelineRuns` can be created in a pending state so that they are not started at all.
+
+## Motivation
+
+The primary motivation for this is to support operators who need to control the
+start up of `PipelineRuns` in clusters that are under heavy load. By allowing
+`PipelineRuns` to be created without actually starting them, one could implement
+a custom controller to implement any `PipelineRun` scheduling policy that is
+transparent to the `PipelineRun` controller and `PipelineRun` creator (e.g.,
+Tekton triggers).
+
+With this feature, platforms can control when `PipelineRuns` are started without
+introducing a bespoke system for storing `PipelineRuns` - potentially adding new
+failure scenarios or UI complexity.
+
+### Goals
+
+* Provide a pending mechanism for `PipelineRuns` that operators can use
+  to control whether or not a `PipelineRun` is running (e.g., "creating new
+  `TaskRuns`").
+
+### Non-Goals
+
+* Support pausing of a running `TaskRun` as that would requiring pausing the
+  Kubernetes pods themselves.
+
+## Requirements
+
+* Provide a mechanism to operators that can be used to mark a `PipelineRun` in a pending
+  state.
+* `PipelineRuns` that are created with the pending setting enabled are not started
+  until the setting is toggled off.
+
+## Proposal
+
+The `PipelineRunSpec` supports a field called `Status` of the type
+`PipelineRunSpecStatus`. Currently the field only supports a single setting,
+`PipelineRunCancelled`. This field will be expanded to support an additional
+setting, `PipelineRunPending`.
+
+When the `Status` field is set to `PipelineRunPending`, the Tekton reconciler
+will set the `Succeeded` `Condition` `Reason` in the `PipelineRun` `Status` to
+`PipelineRunPending`:
+
+```
+status:
+  conditions:
+  - lastTransitionTime: "2020-09-15T18:15:29Z"
+    message: 'PipelineRun is pending.'
+    reason: PipelineRunPending
+    status: "Unknown"
+    type: Succeeded
+```
+
+To start a PipelineRun, the operator should set the `Status` field to empty,
+the Tekton controller will see the change and proceed with the Pipeline as
+normal. Upon starting, the `Succeeded` `Condition` `Reason` will be changed to
+`Running`:
+
+```
+status:
+  conditions:
+  - lastTransitionTime: "2020-09-15T19:46:34Z"
+    message: Not all Steps in the Task have finished executing
+    reason: Running
+    status: Unknown
+    type: Succeeded
+```
+
+If the `PipelineRunPending` setting is set after a PipelineRun has started, the
+validating webhook should reject the update.
+
```

Comment:
it might be worth being explicit around how this state relates to the PipelineRun timing out, i.e. does the timeout start counting down while the PipelineRun is still pending or only once it starts running? (I'm guessing it doesn't start until the PipelineRun actually starts running?)

Reply:
pinging this - id like to be explicit about how this interacts with timeouts in the proposal 🙏

Reply:
Timeouts use the `.status.startTime` field which is set once the PipelineRun has actually started running, not the PipelineRun creation time, so there is no impact. I have added a note. This is ready for merge now.